### PR TITLE
CLSTR-2097:Adding echo loop in ofcontroller to prevent probes from vswitch

### PR DESCRIPTION
**Pull request by nutant-shobhit**
Adding echo loop in ofcontroller to prevent probes from vswitch. ofcontroller is the default controller which runs along with CNC controller. Please refer to https://jira.nutanix.com/browse/CLSTR-2097 for detailed RCA

### CODE REVIEW

File: controller.go [hunk 1]
```diff
@@ -50,8 +51,19 @@ func (c *OFController) ConnectionDown() {
...
+	// echo interval
+	fmt.Println("Controller echo loop started for dp: %v", dp)
+	ticker := time.NewTicker(time.Duration(c.echoInterval) * time.Second)
+	defer ticker.Stop()
+	for range ticker.C {
+		echoReq := ofp13.NewOfpEchoRequest()
+		// ToDo: handle graceful shutdown for stale dps as this
+		// go thread will never exit and may be hung on sendMessage
+		// channel.
+		dp.Send(echoReq)
```
Comment (alolage-nx) [inline controller.go:65]: Given this dp.Send can also block, can we add timed execution for this dp.Send. If this write fails/remains blocked for more than 15 secs, add a panic or at minimum add a log indicating a hang/failure..

File: controller.go [hunk 2]
```diff
@@ -50,8 +51,19 @@ func (c *OFController) ConnectionDown() {
 	// handle connection down
 }
 
-func (c *OFController) sendEchoLoop() {
-	// send echo request forever
+func (c *OFController) sendEchoLoop(dp *Datapath) {
+	// send echo request forever, first tick goes after
+	// echo interval
+	fmt.Println("Controller echo loop started for dp: %v", dp)
```
Comment (alolage-nx) [inline controller.go:57]: can you add similar log when echo loop exits?

File: controller.go (hunk 1 above)
Comment (nutant-shobhit) [reply controller.go:65]: I would say we should just close the channel for which dp.conn has been closed, and there is already a log before that. That will lead to panic in any thread trying to send to a closed channel or blocked on it

File: controller.go (hunk 2 above)
Comment (nutant-shobhit) [reply controller.go:57]: sure

Comment (nutant-shobhit) [reply controller.go:57]: Actually no point adding the log after the loop, it won't get executed as we will panic

Comment (alolage-nx) [reply controller.go:57]: yeah if we panic then no point.

File: controller.go (hunk 1 above)
Comment (alolage-nx) [reply controller.go:65]: you mean add the handling in sendLoop to close channel?

Comment (nutant-shobhit) [reply controller.go:65]: Will be doing improvements as follow up PRs. https://jira.nutanix.com/browse/CLSTR-2144

**Review by alolage-nx — APPROVED** [review]

**Pull request merged**